### DISCUSSION TITLE
canonical paths in output

### DISCUSSION THREAD

**Discussion by hmoffatt** [rizsotto/Bear#425] [Q&A] [answered]
I'm using bear to create a compile_commands.json when compiling my qmake (Qt) project with make.

My problem is that I have a lot of non-canonical paths in the resulting file, of the form `/home/me/path/path/../../src/foo.cpp`.

I want to merge the commands files from multiple projects in to one, but the non-canonical paths means the same file appears multiple times in the output.

**Comment by rizsotto** [comment, marked as answer]
Hey @hmoffatt , can you be explicit which version of Bear are you using?

**Reply by rizsotto** [reply]
There is a new version on `devel` branch, which has a fix for this. @hmoffatt could you try it and tell me if it works for you?

**Reply by hmoffatt** [reply]
I've built it but I'm having some trouble running it;

```
ERROR: ld.so: object '/usr/local/$LIB/bear/libexec.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
```

I did `mkdir build && cd build && cmake .. && make -j4 && sudo make install` ; did I miss something?

Previously I was using the Debian package (3.0.16) and have not built bear from source before.

**Reply by hmoffatt** [reply]
Sorry, read the instructions and tried again with `cmake -DCMAKE_INSTALL_LIBDIR=lib ..` but no change.

**Reply by rizsotto** [reply]
Check out the `INSTALL.md`, it has detailed instructions...

I have added integration test for this, so I am confident it will work. :smile:

**Reply by hmoffatt** [reply]
I've reviewed the INSTALL document and I don't see what I'm doing wrong.

**Reply by rizsotto** [reply]
Ok, I can help if you tell me more about the machine you want it to install. What OS? What distribution? What compiler do you have? Which directory your `libc.so` is in?

Maybe the [troubleshooting](https://github.com/rizsotto/Bear/wiki/Troubleshooting#error-messages-which-appears-with-bear) page will help?

**Reply by hmoffatt** [reply]
Debian 11 (bullseye). cmake says

```
$ cmake -DCMAKE_INSTALL_LIBDIR=lib -DENABLE_UNIT_TESTS=OFF -DENABLE_FUNC_TESTS=OFF  ..                                                 [ 9:02PM]
-- Looking for nlohman_json dependency
-- Looking for nlohman_json dependency -- not found
-- Looking for fmt dependency
-- Looking for fmt dependency -- not found
-- Looking for spdlog dependency
-- Looking for spdlog dependency -- not found
-- Looking for gRPC::grpc++ dependency
-- Checking for modules 'protobuf>=3.11;grpc++>=1.26'
--   No package 'protobuf' found
--   No package 'grpc++' found
-- Looking for gRPC::grpc++ dependency -- not found
-- Configuring done
-- Generating done
-- Build files have been written to: /home/hamish/src/bear/Bear/build
```

After building, BearSource/config.h says SUPPORT_MULTILIB and ```constexpr char DEFAULT_PATH[] = "/usr/local/$LIB/bear/libexec.so";```.

I see this error on the troubleshooting page. I tried setting LIBDIR during cmake but it didn't help. I'll check the Debian packaging.

**Reply by rizsotto** [reply]
Ok, in debian you will need `-DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu`. This is what the `INSTALL.md` was suppose to explain to users. It might be the problem with how it is written. So, please suggest a different wording of the `INSTALL.md` (or the troubleshooting page) which would make it more helpful for you.

**Reply by hmoffatt** [reply]
Thanks, now it builds OK, and I see nice canonical paths in the output.

**Reply by hmoffatt** [reply]
I'm not sure how to improve the instructions. Perhaps the default could be a non-multilib build, as this is where the complexity comes from?

It's confusing that when `CMAKE_INSTALL_LIBDIR=lib` the build doesn't work - the library is in /usr/local/lib, just that the $LIB substitution fails to happen in this case.

**Comment by rizsotto** [comment]
Check out the new `4.0.0` version, where you can select the path formats in the config file.
